## Etapa: 01. Tratamento Dados
#### Código criado por: Eduardo Arthur Bitencourt  | Data: 31/07/2023
##### Github: https://github.com/bitencourt-eduardo  | Linkedin: https://www.linkedin.com/in/bitencourt-eduardo/

In [38]:
# !pip install pandas

In [39]:
#Importar Bibliotecas
import pandas as pd

#### Carregando arquivos FieldPRO

Sensor_FieldPRO, dados transmitidos pelo sensor

In [40]:
# Carregar arquivo de dados transmitidos pelo sensor
name_file = 'Sensor_FieldPRO.csv'
diretorio = '../datasets/{name_file}'

df_sensor = pd.read_csv(diretorio.format(name_file=name_file), sep=',', decimal='.' )
df_sensor.head()

Datetime – utc  air_humidity_100  air_temperature_100  \
0  2020-09-30T23:00:00Z              38.0               31.366   
1  2020-10-01T00:00:00Z               NaN                  NaN   
2  2020-10-01T01:00:00Z              39.0               31.366   
3  2020-10-01T02:00:00Z              39.0               31.322   
4  2020-10-01T03:00:00Z              38.0               31.240   

   atm_pressure_main  num_of_resets  piezo_charge  piezo_temperature  
0               9412              0         45123                 30  
1               9412              0         45025                 31  
2               9419              0         44923                 31  
3               9419              0         44825                 31  
4               9416              0         44728                 31

In [41]:
df_sensor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1711 entries, 0 to 1710
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Datetime – utc       1711 non-null   object 
 1   air_humidity_100     1705 non-null   float64
 2   air_temperature_100  1705 non-null   float64
 3   atm_pressure_main    1711 non-null   int64  
 4   num_of_resets        1711 non-null   int64  
 5   piezo_charge         1711 non-null   int64  
 6   piezo_temperature    1711 non-null   int64  
dtypes: float64(2), int64(4), object(1)
memory usage: 93.7+ KB


* Data e hora estão campo time-stamp ISO-8601 cujo time zone é UTC-0, 
    * contudo não detectado Dtype correto pelo pandas, indicando alguma linha com dado incorreto
* Observado valores nullos para humidade do ar e temperatura

___________________________________________________
Corrigindo formato time-stamp do campo data

In [42]:
df_sensor['datetime'] = pd.to_datetime(df_sensor['Datetime – utc'], format='ISO8601')
df_sensor.head()

Datetime – utc  air_humidity_100  air_temperature_100  \
0  2020-09-30T23:00:00Z              38.0               31.366   
1  2020-10-01T00:00:00Z               NaN                  NaN   
2  2020-10-01T01:00:00Z              39.0               31.366   
3  2020-10-01T02:00:00Z              39.0               31.322   
4  2020-10-01T03:00:00Z              38.0               31.240   

   atm_pressure_main  num_of_resets  piezo_charge  piezo_temperature  \
0               9412              0         45123                 30   
1               9412              0         45025                 31   
2               9419              0         44923                 31   
3               9419              0         44825                 31   
4               9416              0         44728                 31   

                   datetime  
0 2020-09-30 23:00:00+00:00  
1 2020-10-01 00:00:00+00:00  
2 2020-10-01 01:00:00+00:00  
3 2020-10-01 02:00:00+00:00  
4 2020-10-01 03:00:00+00:00

In [43]:
df_sensor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1711 entries, 0 to 1710
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype              
---  ------               --------------  -----              
 0   Datetime – utc       1711 non-null   object             
 1   air_humidity_100     1705 non-null   float64            
 2   air_temperature_100  1705 non-null   float64            
 3   atm_pressure_main    1711 non-null   int64              
 4   num_of_resets        1711 non-null   int64              
 5   piezo_charge         1711 non-null   int64              
 6   piezo_temperature    1711 non-null   int64              
 7   datetime             1711 non-null   datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1), float64(2), int64(4), object(1)
memory usage: 107.1+ KB


In [44]:
df_sensor.describe()

air_humidity_100  air_temperature_100  atm_pressure_main  \
count       1705.000000          1705.000000        1711.000000   
mean          60.238710            24.660940        9432.838106   
std           22.100228             6.174219          26.378333   
min           13.000000            12.360000        9350.000000   
25%           41.000000            19.900000        9416.000000   
50%           61.000000            23.020000        9433.000000   
75%           80.000000            29.400000        9450.000000   
max           99.000000            42.664000        9507.000000   

       num_of_resets  piezo_charge  piezo_temperature  
count    1711.000000   1711.000000        1711.000000  
mean        1.646406  42462.030392          24.910579  
std         1.090382  13038.238287           8.374122  
min         0.000000  16950.000000          11.000000  
25%         1.000000  32020.000000          18.000000  
50%         2.000000  41765.000000          22.000000  
75%         3.000000  53769.000000          32.000000  
max         3.000000  65533.000000          47.000000

_____________________________________________________________________________________________________________________________

#### Carregando arquivos Estação Convencional

Estacao_Convencional, dados transmitidos pela estão metereológica próxima

In [45]:
name_file = 'Estacao_convencional.csv'
diretorio = '../datasets/{name_file}'

df_estacao = pd.read_csv(diretorio.format(name_file=name_file), sep=',', decimal='.' )
df_estacao.head()

data Hora (Brasília)  chuva
0  2020-09-01        00:00:00    0.0
1  2020-09-01        01:00:00    0.0
2  2020-09-01        02:00:00    0.0
3  2020-09-01        03:00:00    0.0
4  2020-09-01        04:00:00    0.0

In [46]:
df_estacao.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2256 entries, 0 to 2255
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   data             2256 non-null   object 
 1   Hora (Brasília)  2256 non-null   object 
 2   chuva            2256 non-null   float64
dtypes: float64(1), object(2)
memory usage: 53.0+ KB


In [47]:
df_estacao.describe()

chuva
count  2256.000000
mean      0.065780
std       0.685743
min       0.000000
25%       0.000000
50%       0.000000
75%       0.000000
max      16.800000

* Nenhum valor nulo observado
* Observados 
    * 2256 registros transmitidos pela estação metereológica
    * 1711 registros transmitidos pela placa FieldPro


* Campos data e hora colunas distintas, e time-zone UTC-3h00 (Brasília).
    * Observar horário de verão para o período: _verificado ano de 2020 não houve horário de verão_
    * Necessário compatibilizar bases de tempo(data/hora)

In [48]:
df_estacao['datetime'] = pd.to_datetime(df_estacao['data'] + ' ' + df_estacao['Hora (Brasília)'] + '-03:00', format='ISO8601')
df_estacao.head()

data Hora (Brasília)  chuva                  datetime
0  2020-09-01        00:00:00    0.0 2020-09-01 00:00:00-03:00
1  2020-09-01        01:00:00    0.0 2020-09-01 01:00:00-03:00
2  2020-09-01        02:00:00    0.0 2020-09-01 02:00:00-03:00
3  2020-09-01        03:00:00    0.0 2020-09-01 03:00:00-03:00
4  2020-09-01        04:00:00    0.0 2020-09-01 04:00:00-03:00

In [49]:
df_estacao.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2256 entries, 0 to 2255
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype                    
---  ------           --------------  -----                    
 0   data             2256 non-null   object                   
 1   Hora (Brasília)  2256 non-null   object                   
 2   chuva            2256 non-null   float64                  
 3   datetime         2256 non-null   datetime64[ns, UTC-03:00]
dtypes: datetime64[ns, UTC-03:00](1), float64(1), object(2)
memory usage: 70.6+ KB


________

* Observando período de transmissão dos dados

In [50]:
df_estacao['datetime'].min()

Timestamp('2020-09-01 00:00:00-0300', tz='UTC-03:00')

In [51]:
df_estacao['datetime'].max()

Timestamp('2020-12-03 23:00:00-0300', tz='UTC-03:00')

In [52]:
df_sensor['datetime'].min()

Timestamp('2020-09-30 23:00:00+0000', tz='UTC')

In [53]:
df_sensor['datetime'].max()

Timestamp('2020-12-11 13:00:00+0000', tz='UTC')

* Período (_não há horário de verão para o ano de 2020_):
    
    * Estação metereológica: 0h00 01/09/2020 a 23h00 03/12/2020 (UTC+3)
    * Sensor FieldPRO: 23h00 30/09/2020 a 13h00 11/12/2020 (UTC)


Realizar merge nos dados compatibilizando data/hora considerando UTC
* DE 23h00 30/09/2020 UTC
* ATÉ 23h00 03/12/2020 UTC-3

In [54]:
df_merged = pd.merge(df_sensor, df_estacao, on='datetime')
df_merged.head()

Datetime – utc  air_humidity_100  air_temperature_100  \
0  2020-09-30T23:00:00Z              38.0               31.366   
1  2020-10-01T00:00:00Z               NaN                  NaN   
2  2020-10-01T01:00:00Z              39.0               31.366   
3  2020-10-01T02:00:00Z              39.0               31.322   
4  2020-10-01T03:00:00Z              38.0               31.240   

   atm_pressure_main  num_of_resets  piezo_charge  piezo_temperature  \
0               9412              0         45123                 30   
1               9412              0         45025                 31   
2               9419              0         44923                 31   
3               9419              0         44825                 31   
4               9416              0         44728                 31   

                   datetime        data Hora (Brasília)  chuva  
0 2020-09-30 23:00:00+00:00  2020-09-30        20:00:00    0.0  
1 2020-10-01 00:00:00+00:00  2020-09-30        21:00:00    0.0  
2 2020-10-01 01:00:00+00:00  2020-09-30        22:00:00    0.0  
3 2020-10-01 02:00:00+00:00  2020-09-30        23:00:00    0.0  
4 2020-10-01 03:00:00+00:00  2020-10-01        00:00:00    0.0

Realizado conferência colunas "Datetime - utc",  "Data" e "Hora (Brasília)" estão compatíveis
* 2020-09-30T23:00:00Z equivale a 2020-09-03 20:00 (horário de Brasília)

In [55]:
#Selecionando colunas relevantes
df_merged = df_merged[['datetime', 'air_humidity_100', 'air_temperature_100','atm_pressure_main','num_of_resets','piezo_charge','piezo_temperature','chuva']]
df_merged.head()

datetime  air_humidity_100  air_temperature_100  \
0 2020-09-30 23:00:00+00:00              38.0               31.366   
1 2020-10-01 00:00:00+00:00               NaN                  NaN   
2 2020-10-01 01:00:00+00:00              39.0               31.366   
3 2020-10-01 02:00:00+00:00              39.0               31.322   
4 2020-10-01 03:00:00+00:00              38.0               31.240   

   atm_pressure_main  num_of_resets  piezo_charge  piezo_temperature  chuva  
0               9412              0         45123                 30    0.0  
1               9412              0         45025                 31    0.0  
2               9419              0         44923                 31    0.0  
3               9419              0         44825                 31    0.0  
4               9416              0         44728                 31    0.0

In [56]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1531 entries, 0 to 1530
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype              
---  ------               --------------  -----              
 0   datetime             1531 non-null   datetime64[ns, UTC]
 1   air_humidity_100     1525 non-null   float64            
 2   air_temperature_100  1525 non-null   float64            
 3   atm_pressure_main    1531 non-null   int64              
 4   num_of_resets        1531 non-null   int64              
 5   piezo_charge         1531 non-null   int64              
 6   piezo_temperature    1531 non-null   int64              
 7   chuva                1531 non-null   float64            
dtypes: datetime64[ns, UTC](1), float64(3), int64(4)
memory usage: 95.8 KB


In [57]:
#Verificando registros com valores nulos
df_merged.loc[df_merged['air_humidity_100'].isnull()]


datetime  air_humidity_100  air_temperature_100  \
1  2020-10-01 00:00:00+00:00               NaN                  NaN   
6  2020-10-01 05:00:00+00:00               NaN                  NaN   
12 2020-10-01 12:00:00+00:00               NaN                  NaN   
16 2020-10-01 16:00:00+00:00               NaN                  NaN   
19 2020-10-01 21:00:00+00:00               NaN                  NaN   
29 2020-10-02 08:00:00+00:00               NaN                  NaN   

    atm_pressure_main  num_of_resets  piezo_charge  piezo_temperature  chuva  
1                9412              0         45025                 31    0.0  
6                9411              0         44537                 30    0.0  
12               9438              0         43874                 30    0.0  
16               9460              0         43259                 30    0.0  
19               9379              0         41914                 37    0.0  
29               9398              0         41083                 19    0.0

Observado:
* Registros nulos de "air_humidity_100" ocorrem ao mesmo tempo quando nulos "air_temperature_100".
* Os registros são nulos sem ocorrência de chuva.

Em virtude do observado optou-se por remover linhas com dados nulos.

In [58]:
df_merged.dropna(subset=['air_humidity_100'], inplace=True)

In [59]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1525 entries, 0 to 1530
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype              
---  ------               --------------  -----              
 0   datetime             1525 non-null   datetime64[ns, UTC]
 1   air_humidity_100     1525 non-null   float64            
 2   air_temperature_100  1525 non-null   float64            
 3   atm_pressure_main    1525 non-null   int64              
 4   num_of_resets        1525 non-null   int64              
 5   piezo_charge         1525 non-null   int64              
 6   piezo_temperature    1525 non-null   int64              
 7   chuva                1525 non-null   float64            
dtypes: datetime64[ns, UTC](1), float64(3), int64(4)
memory usage: 107.2 KB


In [60]:
df_merged.describe()

air_humidity_100  air_temperature_100  atm_pressure_main  \
count       1525.000000          1525.000000        1525.000000   
mean          58.704262            24.776121        9432.478689   
std           21.933831             6.305237          27.272016   
min           13.000000            12.360000        9350.000000   
25%           40.000000            19.900000        9415.000000   
50%           59.000000            23.120000        9433.000000   
75%           78.000000            29.720000        9451.000000   
max           99.000000            42.664000        9507.000000   

       num_of_resets  piezo_charge  piezo_temperature        chuva  
count    1525.000000   1525.000000        1525.000000  1525.000000  
mean        1.495082  44192.804590          25.028197     0.093377  
std         1.040900  12650.914075           8.522773     0.829280  
min         0.000000  16950.000000          11.000000     0.000000  
25%         1.000000  34701.000000          18.000000     0.000000  
50%         1.000000  43968.000000          22.000000     0.000000  
75%         2.000000  54999.000000          32.000000     0.000000  
max         3.000000  65533.000000          47.000000    16.800000

# Exportando dataset tratado e combinado.

In [61]:
df_merged.to_csv('../datasets/df_merged.csv', sep = ';', index=False)